nixos/modules/virtualisation/qubes-dom0.nix

```diff
@@ -0,0 +1,348 @@
+{lib, pkgs, config, ...}:
+let
+  inherit (lib) mkEnableOption mkIf optional mkPackageOption listToAttrs imap1 mkOption submodule concatStringsSep stringAfter;
+  inherit (lib.types) listOf path str enum;
+  inherit (pkgs) qubes-linux-utils qubes-core-qubesdb qubes-core-libvirt qubes-gui-daemon qubes-vmm-xen qemu_qubes qubes-manager runCommand xorg;
+  inherit (pkgs.python3Packages) qubes-core-admin qubes-core-qrexec qubes-desktop-linux-common qubes-core-admin-client qubes-app-linux-usb-proxy;
+
+  cfg = config.virtualisation.qubes;
+
+  symlinkJoin =
+    args_@{ name
+    , paths
+    , preferLocalBuild ? true
+    , allowSubstitutes ? false
+    , stripPrefix ? ""
+    , postBuild ? ""
+    , ...
+    }:
+    let
+      args = removeAttrs args_ [ "name" "postBuild" ]
+        // {
+        inherit preferLocalBuild allowSubstitutes stripPrefix;
+        passAsFile = [ "paths" ];
+      }; # pass the defaults
+    in
+    runCommand name args
+      ''
+        mkdir -p $out
+        for i in $(cat $pathsPath); do
+          ${xorg.lndir}/bin/lndir -silent $i$stripPrefix $out
+        done
+        ${postBuild}
+      '';
+
+  optOutFlags = [
+    "disable-smt"
+    "hostname-dom0"
+  ];
+
+  # Is recommended option enabled?
+  recommended = flag:
+    assert (builtins.elem flag optOutFlags);
+    !(builtins.elem flag cfg.optOutRecommendedConfiguration);
+
+in {
+  # Bug:
+  #   Loopback vchan connection not supported
+  #   FATAL: vchan initialization failed
+  # Solution: Xenstore has no set domid
+  #   It can be set manually using
+  #   `run0 xenstore-write "/domid" 0`
+  #   However it should always be set by default by something,
+  #   and I am not sure how/where.
+  # FIXME: Is this still relevant? I don't remember where I have encountered this error,
+  # but I no longer have this problem.
+
+  # Bug:
+  #   Not enough memory during start of VM
+  # Solution:
+  #   Qubes memory manager overrides max memory of dom0 with
+  #   total memory available, you need to set memory limit for it.
+  #   I have no idea what writes this memory limit in QubesOS, but
+  #   it can be set manually:
+  #   `run0 xenstore-write /local/domain/0/memory/static-max 100000000`
+  #   Note that the standard `xl mem-set` doesn't work for that, as it
+  #   uses /vm/uuid/memory field instead.
+  # FIXME: Permanent solution needed.
```

Comment:
Both of those looks like missing `xen-init-dom0` call (normally a systemd service of this name, calling `/usr/libexec/xen/bin/xen-init-dom0`) - part of xen package

Reply:
The Xen module is enabled [here](https://github.com/NixOS/nixpkgs/pull/341215/files#diff-9abfaa3b06166d20084f02a1d2bda37f9e492a2c7b017b72a37d017c0141171cR298), and that in turn enables the `xen-init-dom0` systemd service [here](https://github.com/NixOS/nixpkgs/blob/c6b2d0f9ade2f273e31b72c2acfef1cc1db752b6/nixos/modules/virtualisation/xen-dom0.nix#L832-L835).

Well, that's what should happen. In theory. It works with vanilla Xen.

Reply:
First problem is fixed even without domid key.
I believe it was caused by hostname being not dom0, thus causing qubes-core-admin-client to misbehave, because it can't find domain according to hostname (qubes expects adminvm to have name == dom0 with no ability to change it, and if dom0 hostname is not dom0 - then it can't detect that current domain is dom0)

This was fixed by changing default dom0 hostname to dom0, and by setting qubesdb /name key to dom0 (Which is used as an alternative method of hostname detection) for the cases where you can't use dom0 as hostname.

Reply:
> qubes expects adminvm to have name == dom0 with no ability to change it

Right... I know it's stupid, but unfortunately it's hardcoded in too many places right now...

Reply:
qubesdb /name workaround works well for everything